File: JavaScript/1696-Jump Game VI.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var maxResult = function(nums, k) {
    stack = [0];
    
    for (let i=1; i<nums.length; i++){
        if (i-k > stack[0]){
            stack.shift();
        }
        nums[i] += nums[stack[0]];
        while (stack.length && nums[i] > nums[stack[stack.length-1]]){
            stack.pop();
        }
        stack.push(i);
    }

    return nums[nums.length-1];
};

File: 1696-Jump Game VI.py
class Solution(object):
    def maxResult(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
# monotone stack
        stack = collections.deque([0])
    
        for i in range(1, len(nums)):
            if i - k > stack[0]:
                stack.popleft()
            nums[i] += nums[stack[0]]
            while stack and nums[i] > nums[stack[-1]]:
                stack.pop()
            stack.append(i)

        return nums[-1]